File: jsonforms.py
"""
Advanced forms library allowing "nesting" one form inside another.
Uses JavaScript and JQuery (1.4+) to return results in JSON format.
"""
from builtins import zip
from past.builtins import basestring
import uuid
from copy import deepcopy
from copy import copy as shallow_copy
try: import json
except ImportError: import simplejson as json

from libedwa.forms import *

class NestedForm(Form):
    """A hybrid of Form and VectorInput"""
    def __init__(self, name, **kwargs):
        self.name = self._name = name
        self.label = kwargs.pop("label", name.replace("_", " ").capitalize())
        self.errors = None
        self.require = kwargs.pop("require", [])
        self._valid = False
        self.kids = None # a list of deep copies of self.children
        if kwargs.pop("required", True): self.require = [not_empty] + self.require
        super(NestedForm, self).__init__(**kwargs)
    def set_data(self, data=[], files=None):
        """
        data    A dictionary mapping HTML names (as strings) to lists of values,
                or a list of such dictionaries, or such a list as a JSON string.
                Bare, non-list values will automatically be wrapped in lists.
                This can either be initial values, or the result of form submission.
                Calling this function replaces any values passed previously or in the constructor,
                and overrides any values passed to individual inputs as initial=...
                In Django, try request.POST['edwa-json']
                In Bottle, try request.forms['edwa-json']
        files   For compatibility with libedwa.Form only.  Do not use.
        
        This can only be called on the top-level nested form.
        Trying to use it directly on nested (child) forms will have no effect,
        because of the way data is propagated to child forms.
        """
        if files: raise ValueError("File uploads not supported!")
        if data is None: data = []
        elif isinstance(data, basestring): data = json.loads(data)
        elif is_scalar(data): data = as_vector(data)
        # Make sure keys are prefixed with the form-wide prefix.
        # Make sure values are lists -- convert bare values into single-item lists.
        self.data = [dict(((k if k.startswith(self.prefix) else self.prefix+k), as_vector(v)) for k, v in datum.items()) for datum in data]
        # Need to re-initialize the subforms
        self.kids = None
    def rawvalues(self):
        return self._values_impl("rawvalues")
    def jsonvalues(self):
        # Errors will only be present if values() or validate() was called.
        return self._values_impl("jsonvalues")
    def values(self):
        # If we've already computed our value, just return it.
        if hasattr(self, "_value"):
            return self._value
        # Enter validation mode. Start accumulating error messages.
        self.errors = []
        self._valid = True
        # Do not escape error messages here: they will be escaped before output.
        # This means they're still readable for e.g. console output.
        try:
            self._value = self._values_impl("values")
        except Exception as ex:
            self._value = []
            self.errors.append(to_unicode(ex))
        # If coerced successfully, apply the 'require' criteria (if any).
        for requirement in self.require:
            try: err = requirement(self._value) # add kwargs here as needed!
            except TypeError: err = requirement(self._value) # old-style validator doesn't take **kwargs
            if err: self.errors.append(to_unicode(err))
        if self.errors:
            self._value = []
            self._valid = False
        return self._value
    def _values_impl(self, mode):
        if self.kids is None:
            # Break references between children and self to reduce scope of deep copying:
            old_forms = [(child, getattr(child, "form", None)) for child in self.children]
            for child in self.children:
                if hasattr(child, "form"): child.form = None
            # Make a deep copy of self.children for each set of data we have:
            self.kids = []
            for datum in self.data:
                #new_kids = deepcopy(self.children)
                new_kids = [shallow_copy(child) for child in self.children]
                for kid in new_kids:
                    # Otherwise, kid.form points to a new, copied form object:
                    if hasattr(kid, "form"): kid.form = self
                    # This is used by the validation machinery for inter-Input dependencies:
                    kid.peers = new_kids # since kid.form.children refers to the unused, prototypical NestedForm.children
                    # This is the proper time to set data, so it isn't re-set unnecessarily.
                    if isinstance(kid, NestedForm): kid.set_data(datum.get(kid._name))
                self.kids.append(new_kids)
            # Restore links between (prototypical) children and self:
            for child, old_form in old_forms:
                if hasattr(child, "form"): child.form = old_form
        vals = []
        all_data = self.data
        try:
            # Define mode of data retrieval and error handling / validation
            if mode == "rawvalues":
                def prop(child): return child.rawvalue
            elif mode == "jsonvalues":
                def prop(child):
                    p = (child.jsonvalues() if hasattr(child, "jsonvalues") else child.rawvalue)
                    if child.errors and not isinstance(child, NestedForm): p = {"value":p, "errors":child.errors}
                    return p
            elif mode == "values":
                def prop(child):
                    self._valid &= child.validate()
                    return child.value
            else: assert False, "Unknown mode of operation!"
            # Cycle through each dataset in the array and process it.
            # Each dataset is processed with a clean copy of the child fields.
            for datum, children in zip(all_data, self.kids):
                self.data = datum
                vals.append(dict((child._name, prop(child)) for child in children))
        finally:
            self.data = all_data
        if mode == "jsonvalues" and self.errors:
            vals = {"value":vals, "errors":self.errors}
        return vals
    def validate(self):
        self.value
        return self._valid
    @property
    def rawvalue(self):
        return self.rawvalues()
    @property
    def value(self):
        return list(self.values())

def make_template(nform, extras, indent=""):
    """Returns as JavaScript object as a string, starting with "{" and ending with "}"."""
    lines = []
    j = json.dumps
    for child in nform:
        if isinstance(child, NestedForm):
            lines.append("%s%s: edwa.jsonforms.make%s(%s, %s, %s, %s)" % (
                indent, j(child._name), child.__class__.__name__, j(child._name), j(child.label), make_template(child, extras, indent+"    "), j(extras)))
        else:
            lines.append("%s%s: edwa.jsonforms.make_input(%s, %s, %s, %s)" % (
                indent, j(child._name), j(child._name), j(child.label), j(child.help_text or ""), j(child.html())))
    return "{\n" + ",\n".join(lines) + "\n" + indent + "}";

def make_html(nform, extras={}, input_name="edwa-json"):
    """Makes a SCRIPT tag and the start of the FORM, still needs a submit button and a closing tag.
    JQuery and jsonforms.js should have already been included in the HEAD."""
    form_id = "%s%s_form" % (nform.id_prefix, nform.prefix)
    form_html = nform.html()
    template = make_template(nform, extras, "    ")
    data = json.dumps(nform.jsonvalues(), indent=4)
    return """<script type="text/javascript">
jQuery(document).ready(function() {
var template = %(template)s;

var data = %(data)s;

jQuery("#%(form_id)s").prepend(edwa.jsonforms.make_jsonform(data, template)).submit(function() {
    // Programatic changes (this.value = ...) to input elements do not generate change events.
    // Thus, for safety, we trigger one manually before submit.  :input matches textareas, selects, checkboxes, etc.
    jQuery(":input").change();
    // JSON.stringify() can be obtained from the "json2" library if the browser does not provide it.
    jQuery("input[name=%(input_name)s]", this)[0].value = JSON.stringify(data);
});
});
</script>
%(form_html)s
<input type="hidden" name="%(input_name)s">
""" % vars()

def path_to_jsonforms_js():
    import os.path as p
    return p.abspath(p.join(p.dirname(__file__), "jsonforms.js"))

def test_me():
    pet_form = NestedForm("pets", required=False)
    pet_form += TextInput(pet_form, "name")
    pet_form += Select(pet_form, "species", choices=("perro", "gato", "pez", ("hamster", raw("h&aacute;mster"))))#, multiple=True)
    
    dep_form = NestedForm("dependents")
    dep_form += TextInput(dep_form, "name")
    dep_form += TextInput(dep_form, "age")
    dep_form += pet_form
    
    form = NestedForm("x")
    form += TextInput(form, "name")
    form += TextInput(form, "phone")
    form += CheckboxInput(form, "call_me")
    form += dep_form
    
    data = {
        "name": "Juan Carlos Doe",
        "phone": "",#919-555-1234",
        "call_me": True,
        "dependents": [
            { "name": "Maria", "age": "5" },
            { "name": "Ana", "age": "3" },
            { "name": "Juanita", "age": "1" },
            { "pets": [{"species": "pez"}] } # an empty slot at the end of the list, with an empty slot for pets!
        ]
    }
    form.set_data(data)
    #print "RAW", form.rawvalues()
    #print "VAL", form.values()
    #print "HTML", make_html(form)
    return form

File: jsonforms.js
/*
- add CSS classes, etc. to the table elements
- add support for other kinds of form elements (except file upload?)
- add a display-only view and allow toggling between edit and viewing mode
*/
if(!edwa) var edwa = {css_prefix: "edwa-"};
(function($) {
edwa.jsonforms = {
    next_id: 10000,
    // Allow each bit of initial data to be a simple value
    // or an object with attributes "value" and "errors".
    get_value: function(x) {
        if($.isPlainObject(x)) return x.value;
        else return x;
    },
    get_errors: function(x) {
        if($.isPlainObject(x)) return x.errors || [];
        else return [];
    },
    // Generate a simple label and corresponding HTML <INPUT> for one form field.
    // Actually, generate a function that will do that, when called with initial data.
    // data will be updated when the form value changes.
    make_input: function(name, label_text, help_text, input_str) {
        function handler(data) {
            var input = $(input_str);
            var input_id = "id_edwa_"+edwa.jsonforms.next_id;
            edwa.jsonforms.next_id += 1;
            input[0].id = input_id;
            var label_th = $("<th align='right'><div class='"+edwa.css_prefix+"label'><label for='"+input_id+"'>"+label_text+"</label></div></th>");
            var td = $("<td></td>");
            var input_div = $("<div class='"+edwa.css_prefix+"input'></div>");
            input_div.append(input);
            td.append(input_div);
            var msg_div = $("<div class='"+edwa.css_prefix+"msgs'></div>");
            if(help_text) {
                var help_div = $("<div class='"+edwa.css_prefix+"msgs'>"+help_text+"</div>");
                msg_div.append(help_div);
            }
            if(name in data) {
                // In Firefox, type is "select-one" or "select-multiple".
                if(/^select/.test(input[0].type)) {
                    // Make sure each value is in string format
                    var values = $.map(edwa.jsonforms.get_value(data[name]), function(elem, idx) { return ""+elem; });
                    // Set the selected status of each option in the select
                    $("option", input).each(function() { this.selected = ($.inArray(this.value, values) != -1); });
                } else if(input[0].type == "checkbox" || input[0].type == "radio") {
                    input[0].checked = edwa.jsonforms.get_value(data[name]);
                } else {
                    input[0].value = edwa.jsonforms.get_value(data[name]);
                }
                var errs = edwa.jsonforms.get_errors(data[name]);
                if(errs.length) {
                    var err_ul = $("<ul></ul>");
                    for(var ii in errs) {
                        err_ul.append($("<li>"+errs[ii]+"</li>"));
                    }
                    var err_div = $("<div class='"+edwa.css_prefix+"error'></div>");
                    err_div.append(err_ul);
                    msg_div.append(err_div);
                }
                // Once we've displayed the errors initially, simplify the structure of the JSON again.
                data[name] = edwa.jsonforms.get_value(data[name]) || "";
            }
            td.append(msg_div);
            input.change(function() {
                if(/^select/.test(this.type)) {
                    // Iterate through each choice, adding selected ones to the list of values.
                    var vals = [];
                    $("option", this).each(function() {
                        if(this.selected) vals.push(this.value);
                    });
                    data[name] = vals;
                }
                else if(this.type == "checkbox" || this.type == "radio") data[name] = this.checked;
                else data[name] = this.value;
            }).change(); // need to trigger fake change to sync single SELECTs that start with empty value
            var tr = $("<tr valign='top'></tr>");
            if(input[0].type != "hidden") tr.append(label_th);
            else tr.append("<th align='right'><div class='"+edwa.css_prefix+"label'></th>");
            tr.append(td);
            return tr;
        }
        return handler;
    },
    // Generate a list of nested forms, where one can add and remove items from the list.
    // Actually, generate a function that will do that, when called with initial data.
    // all_data will be updated when the form values change.
    makeNestedForm: function(name, label, template, extras) {
        var extras = $.extend({
            allow_add: true,
            allow_remove: true,
            add_icon: "(+)",
            remove_icon: "(-)",
            _return_td: false
        }, extras);
        function handler(all_data) {
            if(!(name in all_data)) all_data[name] = [];
            // Extract the errors, if any, then simplify the JSON structure.
            var errs = edwa.jsonforms.get_errors(all_data[name]);
            var data_list = edwa.jsonforms.get_value(all_data[name]) || [];
            all_data[name] = data_list;
            var td = $("<td></td>")
            var input_div = $("<div class='"+edwa.css_prefix+"input'></div>");
            td.append(input_div);
            // Make a <TABLE> containing one complete set of form fields,
            // along with a link to remove that set of fields.
            function make_table(data, allow_remove) {
                var table = $("<table border='0' cellspacing='0' cellpadding='0' class='"+edwa.css_prefix+"nform'></table>");
                var nrows = 0;
                for(var fieldname in template) {
                    var row = template[fieldname](data);
                    table.append(row);
                    nrows += 1;
                }
                var remove_this = $("<a href='' class='"+edwa.css_prefix+"rm1'>"+extras.remove_icon+"</a>").click(function() {
                    var ii = $.inArray(data, data_list); // == indexOf
                    if(ii < 0 && console) console.log("Whoa -- can't find form item to remove it!");
                    data_list.splice(ii, 1);
                    table.remove();
                    return false;
                });
                if(allow_remove) {
                    $("tr", table).first().prepend($("<td rowspan='"+nrows+"'></td>").append(remove_this));
                }
                // For any SELECTs that are marked with CSS class "edwa-new-and-used",
                // hide the other fields in that subform UNLESS the value is 0.
                // This facilitates either choosing between existing objects
                // or creating a new object. (Editing an existing object is not supported.)
                $("."+edwa.css_prefix+"new-and-used", table).bind("change", function() {
                    if(this.value == 0) {
                        $("tr", $(this).closest("table")).show();
                    } else {
                        var my_tr = $(this).closest("tr");
                        $("tr", $(this).closest("table")).not(my_tr).hide();
                    }
                }).change(); // initial sync
                return table;
            }
            // For each item in the list of initial values, construct a <TABLE>
            // with a full complement of fields.
            for(ii in data_list) {
                var data = data_list[ii];
                // Fields created from initialization data may or may not be removable.
                input_div.append(make_table(data, extras.allow_remove));
            }
            // Create a link that will add a (blank) item to the list.
            var add_one = $("<a href='' class='"+edwa.css_prefix+"add1'>"+extras.add_icon+"</a>").click(function() {
                var data = {};
                data_list.push(data);
                // Newly created entries are always removable before form is submitted.
                $(this).before(make_table(data, true));
                return false;
            });
            if(extras.allow_add) {
                input_div.append(add_one); // below list of forms
            }
            if(errs.length) {
                var err_ul = $("<ul></ul>");
                for(var ii in errs) {
                    err_ul.append($("<li>"+errs[ii]+"</li>"));
                }
                var err_div = $("<div class='"+edwa.css_prefix+"error'></div>");
                err_div.append(err_ul);
                var msg_div = $("<div class='"+edwa.css_prefix+"msgs'></div>");
                msg_div.append(err_div);
                td.append(msg_div);
            }
            var label_td = $("<th align='right'><div class='"+edwa.css_prefix+"label'> "+label+" </th>");
            var row = $("<tr valign='top'></tr>").append(label_td).append(td);
            if(extras._return_td) return td;
            else return row;
        }
        return handler;
    },
    // Re-use machinery from above to create layout for outer object.
    // To be proper JSON, this must be a single object -- no adding/removing from a list!
    make_jsonform: function(data, template, extras) {
        var extras = $.extend(extras, {
            allow_add: false,
            allow_remove: false,
            _return_td: true
        });
        var data = edwa.jsonforms.get_value(data);
        if(data.length == 0) data.push( {} );
        var td = edwa.jsonforms.makeNestedForm("x", "x", template, extras)({x:data});
        return td.children().first().children(); // all DOM elements in the DIV in the second top-level TD
    }
};
})(jQuery);
